### PR TITLE
feat(discovery): surface a stable USB physical-location identifier

### PR DESCRIPTION
**Pull request by tylerkron**
## Summary

Closes #285. Adds `IUsbLocationProvider` — a stable USB physical-location key (Windows: WMI `DEVPKEY_Device_LocationInfo`, other platforms: no-op fallback) — and wires `IDeviceInfo.LocationKey` through `SerialDeviceFinder` and `HidDeviceFinder`, so the same physical device can be correlated across serial ⇄ HID-bootloader mode transitions and multiple identical bootloaders (same VID/PID, no serial number) can be disambiguated.

- `IUsbLocationProvider` / `WindowsUsbLocationProvider` / `NullUsbLocationProvider` / `UsbLocationProviderFactory` — mirrors the existing `IUsbPortDescriptorProvider` pattern. The Windows implementation resolves a COM port name or HID device interface path to `DEVPKEY_Device_LocationInfo` via WMI's `GetDeviceProperties` method call, matching the issue's verified reference (`Get-PnpDeviceProperty ... DEVPKEY_Device_LocationInfo`).
- `HidDevicePathParser.ParseInstanceId` — extracts a PnP instance ID from a HID device interface path via documented string manipulation (no new `SetupDi`/`Cfgmgr32` P/Invoke surface), kept in its own untagged type so the pure logic is unit-testable without a Windows-only WMI dependency.
- `IDeviceInfo`/`DeviceInfo.LocationKey`, populated by both finders via an injectable provider (constructor param, same shape as the existing descriptor-provider seam).
- `FirmwareUpdateService`/`IFirmwareUpdateService` gain a `targetLocationKey` parameter alongside `targetDevicePath` (path wins if both are given). `WaitForBootloaderDeviceAsync` resolves each candidate bootloader's location and matches on it, letting a caller target the bootloader a device reboots *into* using the location it observed *before* the reboot — closing the first-match auto-update limitation documented in daqifi-desktop's `FirmwareUpdateCoordinator.cs`.

## What's verified vs. not

- Unit tests cover the platform-independent logic (`NullUsbLocationProvider`, `HidDevicePathParser` parsing edge cases) and the discovery/firmware-service wiring via hand-written fakes, following this repo's existing no-mocking-library convention.
- Ran the example CLI against this branch's `Daqifi.Core` on real hardware (macOS bench rig): `--discover-serial` and a connect+stream smoke test both succeeded, confirming the new `LocationKey` code path in `SerialDeviceFinder` doesn't regress discovery (resolves to `null` via the macOS no-op fallback, as expected).
- **Not verified**: the real Windows WMI calls (`GetDeviceProperties`) — no Windows machine in this environment, and CI is `ubuntu-latest` only, so only the pure parsing/fallback logic has real test coverage (same limitation the pre-existing `WindowsUsbPortDescriptorProvider` has).
- **Not verified**: the acceptance criterion that the location key is identical for the same physical device in serial vs. HID mode — needs a Windows bench test with real hardware.

## Test plan

- [x] `dotnet build Daqifi.Core.sln` — clean, 0 warnings/errors (repo has `TreatWarningsAsErrors`)
- [x] `dotnet test Daqifi.Core.sln` — 1403 passed, 2 pre-existing hardware-only skips, 0 failed
- [x] Real-hardware serial discovery + connect/stream smoke test (see above)
- [ ] Manual Windows bench verification of `WindowsUsbLocationProvider` and the serial↔HID location-key stability assumption (follow-up, not done in this PR)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Daqifi.Core.Tests/Device/Discovery/SerialDeviceFinderTests.cs [hunk 1]
```diff
@@ -111,6 +111,22 @@ public void SerialDeviceFinder_CustomBaudRate_AcceptsCustomBaudRate()
...
+        // simulate — the same limitation that already applies to every other field mapped
+        // from statusMessage there (SerialNumber, FirmwareVersion, etc.). This test locks in
+        // the constructor wiring; end-to-end LocationKey population is exercised on real
+        // hardware, not here.
+        var fakeProvider = new RecordingUsbLocationProvider(_ => "Port_#0001.Hub_#0001");
+
+        using var finder = new SerialDeviceFinder(9600, usbPortDescriptorProvider: null, usbLocationProvider: fakeProvider);
+
+        Assert.NotNull(finder);
+    }
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply src/Daqifi.Core.Tests/Device/Discovery/SerialDeviceFinderTests.cs:122]: Agreed the original test only proved constructor wiring, not actual `LocationKey` population. Implemented differently than the suggested fix though — rather than adding an injectable serial-port/stream transport seam to `SerialDeviceFinder` (a much larger surface-area change, and risky to retrofit onto a class that's already validated on real hardware), I extracted the `DeviceInfo`-construction logic into a new `internal static SerialDeviceFinder.BuildDeviceInfo(DaqifiOutMessage, string portName, string? locationKey)` method — mirroring the existing `ConvertDeviceType` pattern in the same file, which is already tested this way.

`DaqifiOutMessage` is a plain protobuf-generated class with a public constructor and settable fields, so it can be constructed directly in a test with no real serial port involved. New tests (`BuildDeviceInfo_MapsStatusMessageAndLocationKey`, `BuildDeviceInfo_WithNullLocationKey_LeavesLocationKeyNull`, `BuildDeviceInfo_WithBlankDevicePn_FallsBackToPortNameAsName`) now exercise the real mapping logic, including `LocationKey`, without any fake transport.

`TryGetDeviceInfoAsync` now resolves `locationKey` (with its own try/catch — see the reply on the "Discovery fails on provider throw" item) and delegates to `BuildDeviceInfo`. Commit 8bb3abd.

File: src/Daqifi.Core/Device/Discovery/IDeviceInfo.cs [hunk 2]
```diff
@@ -69,6 +69,16 @@ public interface IDeviceInfo
...
+
+    /// <summary>
+    /// Gets the USB physical-location key (e.g. <c>Port_#0001.Hub_#0001</c>), or null if it
+    /// could not be resolved. Stable for a given physical USB port across a device's
+    /// transitions between transports (e.g. serial app mode ⇄ HID bootloader mode) and
+    /// re-enumerations, so it can be used to correlate the same physical unit and disambiguate
+    /// multiple identical devices (same VID/PID, no serial number). Resolved via
+    /// <see cref="IUsbLocationProvider"/>; Windows-only in v1, always null elsewhere.
+    /// </summary>
+    string? LocationKey { get; }
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply src/Daqifi.Core/Device/Discovery/IDeviceInfo.cs]: Agreed this was a real bug — good catch. Went with a different fix than the suggested `ILocationKeyDeviceInfo : IDeviceInfo` split interface though: `IDeviceInfo.LocationKey` is now a **default interface property** (`string? LocationKey => null;`) instead of a plain abstract member.

This is the same technique this repo already uses on `IFirmwareUpdateService` for its targeted-update overloads (see that interface's own doc comments: "Added as a default interface method so existing implementers... are unaffected"). An external `IDeviceInfo` implementer compiled before this member existed now inherits the default (`null`, meaning "not resolved") without recompiling — no `TypeLoadException` risk — and `DeviceInfo` overrides it with a real stored value exactly as before.

I think this is a better fix than the split-interface for callers: with `ILocationKeyDeviceInfo`, every consumer that wants `LocationKey` off an `IDeviceInfo` they didn't create themselves has to pattern-match/cast and handle the case where the concrete type doesn't implement it. With the default-property approach, `device.LocationKey` just works on any `IDeviceInfo`, defaulting sensibly to null — which is also exactly the property's documented "null if not resolved" contract, so the default isn't just a compatibility shim, it's the semantically correct fallback.

Proof it actually fixes the compat issue: I reverted the `LocationKey` addition to the test project's own hand-written `TestDeviceInfo : IDeviceInfo` stub (`DaqifiDeviceFactoryTests.cs`) and it still compiles and passes, unmodified, against the new interface — same situation an external implementer would be in. Commit 8bb3abd.

File: docs/DEVICE_INTERFACES.md [hunk 3]
```diff
@@ -125,6 +125,32 @@ if (devices.Any())
...
+// ...device reboots into bootloader mode via ForceBootloader...
+
+// Target the bootloader that came from the SAME physical port, even though its
+// HID device path didn't exist until after the reboot.
+await firmwareUpdateService.UpdateFirmwareAsync(
+    device, hexFilePath, progress: null, targetDevicePath: null, targetLocationKey: targetLocation);
+```
+
+`LocationKey` is resolved via `IUsbLocationProvider` and is Windows-only in v1 (Linux/macOS
+resolve to `null`, same as `IUsbPortDescriptorProvider`'s cross-platform fallback pattern).
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply docs/DEVICE_INTERFACES.md:152]: Fair — the doc presented the serial↔HID stability claim as settled fact, but it hasn't actually been confirmed on Windows hardware (the environment this PR was built in has no Windows machine, and CI is `ubuntu-latest` only). Added an explicit "Verification status" callout right under the claim in `DEVICE_INTERFACES.md` stating this is unverified and needs a Windows bench check before being relied on for anything safety-critical, and noting the same gap applies to `WindowsUsbLocationProvider`'s WMI query path. I didn't fabricate verification steps/results since none exist yet — the honest fix here is disclosure, not a false claim of completeness. Commit 8bb3abd.
